File: daily_python/3583_Count_Special_Triplets.py
import os
import sys
from typing import List
from collections import defaultdict

class Solution:
    def specialTriplets(self, nums: List[int]) -> int:
        left = defaultdict(int)
        right = defaultdict(int)
        ans = 0
        mod = 10 ** 9 + 7
        for i in nums:
            right[i] += 1
        for i in nums:
            right[i] -= 1
            l = left[i * 2]
            r = right[i * 2]
            ans = (ans + l * r) % mod
            left[i] += 1
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            dataset.append(nums)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums in dataset:
        results.append(solution.specialTriplets(nums))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2025/3583_Count_Special_Triplets.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int specialTriplets(vector<int>& nums) {
        int n = nums.size();
        long long ans = 0;
        int mod = 1e9 + 7;
        unordered_map<int,int> left;
        unordered_map<int,int> right;
        for (int i : nums) right[i]++;
        for (int i = 0; i < n; i++){
            int curr = nums[i];
            right[curr]--;
            int l = left[curr * 2];
            int r = right[curr * 2];
            ans = (ans + 1LL * l * r) % mod;
            left[curr]++;
        }
        return ans;
    }
};

int main(){
    vector<int> nums = {6,3,6};
    // vector<int> nums = {0,1,0,0};
    // vector<int> nums = {8,4,2,8,4};
    Solution S;
    cout << S.specialTriplets(nums) << endl;
    return 0;
}